activemodel/lib/active_model/attributes.rb

```diff
@@ -1,24 +1,31 @@
 # frozen_string_literal: true
 
+require "active_support/core_ext/object/deep_dup"
 require "active_model/type"
+require "active_model/attribute_set"
+require "active_model/attribute/user_provided_default"
 
 module ActiveModel
   module Attributes #:nodoc:
     extend ActiveSupport::Concern
     include ActiveModel::AttributeMethods
-    include ActiveModel::Dirty
 
     included do
       attribute_method_suffix "="
       class_attribute :attribute_types, :_default_attributes, instance_accessor: false
-      self.attribute_types = {}
-      self._default_attributes = {}
+      self.attribute_types = Hash.new(Type.default_value)
+      self._default_attributes = AttributeSet.new({})
     end
 
     module ClassMethods
-      def attribute(name, cast_type = Type::Value.new, **options)
-        self.attribute_types = attribute_types.merge(name.to_s => cast_type)
-        self._default_attributes = _default_attributes.merge(name.to_s => options[:default])
+      def attribute(name, type = Type::Value.new, **options)
+        name = name.to_s
+        if type.is_a?(Symbol)
+          type = ActiveModel::Type.lookup(type, **options.except(:default))
+        end
+        self.attribute_types = attribute_types.deep_dup
```

Comment:
Is there a reason to `deep_dup` here instead of the merge we were doing before?

Reply:
I'll try going back to a merge.  It's been through a lot of intermediate stages, and it might not actually be necessary now.